### PR TITLE
Fixes detach at silences by trimming unnecessary silence #2312

### PR DESCRIPTION
**Pull request by gpapp**
HandleSplit in 3.x.x is using the new Trim features, instead of Clear. When using the noise gate and detach silence for tracks, this results in full copies of the orginal track, with enormous files.

When HandleSplit is modified to use Clear, the operation falls back to the intended way.

As a side effect, when using this modifications, there are no unintentional copies of invisible data are created, e.g. spliting a clip actually splits, does not create copies.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [X] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by LWinterberg** [comment]
Fwiw: #2051 has become a much larger project to increase performance in general (with #2065, #2087 and #2121) going much beyond the initial detach at silences issue.

**Comment by gpapp** [comment]
Reworked the modification to only affect the Detach silence function

**Comment by LWinterberg** [comment]
looks good, thanks!

File: src/WaveTrack.cpp [hunk 1]
```diff
@@ -1694,12 +1695,59 @@ void WaveTrack::Disjoin(double t0, double t1)
...
+   for (unsigned int i = 0; i < clipRegions.size(); i++)
+   {
+      const ClipRegion& region = clipRegions.at(i);
+      // There is no segment to be kept on the left
+      if (region.clip->GetPlayStartSample() == region.start)
+      {
+         // the region spans to the end of the clip, so we can just delete it
+         if (region.clip->GetPlayEndSample() == region.end) {
+            WaveClipHolder toRemove = RemoveAndReturnClip(region.clip.get());
+            toRemove.~shared_ptr();
```
Comment (ksoze95) [inline src/WaveTrack.cpp:1714]: No need to explicitly call the destructor.  The destructor will be called when toRemove goes out of scope.

File: src/WaveTrack.cpp [hunk 2]
```diff
@@ -1694,12 +1695,59 @@ void WaveTrack::Disjoin(double t0, double t1)
...
+
+      // lookahead to see if we are done with this clip
+      if ((i == clipRegions.size() - 1) || (region.clip!=clipRegions.at(i+1).clip)) {
+         // the region spans to the end of the clip, so we can just delete it
+         if (region.clip->GetPlayEndSample() != region.end) {
+            newClip = CopyToNewClip(region.clip, region.end, region.clip->GetPlayEndSample());
+         }
+         //remove the remainder of the clip
+         WaveClipHolder toRemove = RemoveAndReturnClip(region.clip.get());
+         toRemove.~shared_ptr();
```
Comment (ksoze95) [inline src/WaveTrack.cpp:1729]: Same as above.  Not needed.

File: src/WaveTrack.cpp (hunk 1 above)
Comment (gpapp) [reply src/WaveTrack.cpp]: Removed explicit destructor call

File: src/WaveTrack.cpp (hunk 2 above)
Comment (gpapp) [reply src/WaveTrack.cpp]: Removed explicit destructor call

**Review by vsverchinsky — CHANGES_REQUESTED** [review]
New clips should reference existing blocks instead of duplicating samples.

File: .gitignore [hunk 3]
```diff
@@ -171,3 +171,4 @@ appimagetool/*
 .build.arm64/
 .conan/
 .debug/
+/out/build/x64-Debug
```
Comment (Paul-Licameli) [inline .gitignore:174]: This is extraneous to the fix

**Comment by LWinterberg** [comment]
Hello! We just [switched master to 4.0.0](https://forum.audacityteam.org/t/master-has-been-switched-to-4-0-0-development/120444) which makes this PR incompatible and/or obsolete as we're switching UI framework and reorganizing large parts of the codebase. You're very welcome to make these kinds of contributions in the future though, after things stabilize a little bit.

**Pull request closed without merging**